src/features/expressions/expression-functions.ts

```diff
@@ -0,0 +1,605 @@
+        // Expressions will run before the layout is fully loaded, so we might not have all the components available
+        // yet. If that's the case, silently ignore this expression.
+        return null;
+      }
+
+      const dataModelBindings = closest
+        ? this.dataSources.nodeDataSelector((picker) => picker(closest)?.layout.dataModelBindings, [closest])
+        : undefined;
+
+      const simpleBinding =
+        dataModelBindings && 'simpleBinding' in dataModelBindings ? dataModelBindings.simpleBinding : undefined;
+      if (closest && simpleBinding) {
+        if (this.dataSources.isHiddenSelector(closest)) {
+          return null;
+        }
+
+        return pickSimpleValue(simpleBinding, this.dataSources.formDataSelector);
+      }
+
+      // Expressions can technically be used without having all the layouts available, which might lead to unexpected
+      // results. We should note this in the error message, so we know the reason we couldn't find the component.
+      const hasAllLayouts = node instanceof LayoutPage ? !!node.layoutSet : !!node.page.layoutSet;
+      throw new ExprRuntimeError(
+        this.expr,
+        this.path,
+        hasAllLayouts
+          ? `Unable to find component with identifier ${id} or it does not have a simpleBinding`
+          : `Unable to find component with identifier ${id} in the current layout or it does not have a simpleBinding`,
+      );
+    },
+    args: [ExprVal.String] as const,
+    returns: ExprVal.Any,
+  }),
+  dataModel: defineFunc({
+    impl(path): any {
+      if (path === null) {
+        throw new ExprRuntimeError(this.expr, this.path, `Cannot lookup dataModel null`);
+      }
+
+      const node = ensureNode(this.node);
+      if (node instanceof BaseLayoutNode) {
+        const newPath = this.dataSources.transposeSelector(node as LayoutNode, path);
+        return pickSimpleValue(newPath, this.dataSources.formDataSelector);
+      }
+
+      // No need to transpose the data model according to the location inside a repeating group when the context is
+      // a LayoutPage (i.e., when we're resolving an expression directly on the layout definition).
+      return pickSimpleValue(path, this.dataSources.formDataSelector);
+    },
+    args: [ExprVal.String] as const,
+    returns: ExprVal.Any,
+  }),
+  externalApi: defineFunc({
+    impl(externalApiId, path): string {
+      if (typeof externalApiId !== 'string' || typeof path !== 'string') {
+        throw new ExprRuntimeError(this.expr, this.path, `Expected string arguments`);
+      }
+
+      const externalApiData = this.dataSources.externalApis[externalApiId]?.data;
+
+      const res =
+        path && externalApiData && typeof externalApiData === 'object'
+          ? dot.pick(path, externalApiData)
+          : externalApiData;
+
+      if (!res) {
+        return ''; // Print error?
+      }
+
+      if (typeof res === 'object') {
+        return JSON.stringify(res);
```

Comment:
I would prefer returning null in that case. We want to add list and object support in the future (#1179), and it's best if app developers start getting used to this returning a JSON string they could try to substring their way out of, which we'll later swap out with real object/arrays... :scream: Currently, `dataModel` lookups also returns null when an object or array is found.

Reply:
Makes sense, and probably results in less maintenance too.